examples/variational_autoencoder/variational_autoencoder.py

```diff
@@ -0,0 +1,326 @@
+
+    # Create VAE model
+    print("Building model and compiling functions...")
+    print("L = {}, z_dim = {}, n_hid = {}, binary={}".format(L, z_dim, n_hid, binary))
+    x_dim = width * height
+    l_z_mu, l_z_ls, l_x_mu_list, l_x_ls_list, l_x_list, l_x = \
+           build_vae(input_var, L=L, binary=binary, z_dim=z_dim, n_hid=n_hid)
+
+    def build_loss(deterministic, withNoise=True):
+        layer_outputs = nn.layers.get_output([l_z_mu, l_z_ls] + l_x_mu_list + l_x_ls_list
+                + l_x_list, deterministic=deterministic)
+        z_mu =  layer_outputs[0]
+        z_ls =  layer_outputs[1]
+        x_mu =  [] if binary else layer_outputs[2:2+L]
+        x_ls =  [] if binary else layer_outputs[2+L:2+2*L]
+        x_list =  layer_outputs[2:2+L] if binary else layer_outputs[2+2*L:2+3*L]
+        # Loss expression has two parts as specified in [1]
+        # kl_div = KL divergence between p_theta(z) and p(z|x)
+        # - divergence between prior distr and approx posterior of z given x
+        # - or how likely we are to see this z when accounting for Gaussian prior
+        # logpxz = log p(x|z)
+        # - log-likelihood of x given z
+        # - in binary case logpxz = cross-entropy
+        # - in continuous case, is log-likelihood of seeing the target x under the
+        #   Gaussian distribution parameterized by dec_mu, sigma = exp(dec_logsigma)
+        kl_div = 0.5 * T.sum(1 + 2*z_ls - T.sqr(z_mu) - T.exp(2 * z_ls))
+        if binary:
+            logpxz = sum(nn.objectives.binary_crossentropy(x,
+                input_var.flatten(2)).sum() for x in x_list) * (-1./L)
+            prediction = nn.layers.get_output(l_x, deterministic=deterministic,
+                    withNoise=withNoise)
+        else:
+            logpxz = sum(log_likelihood(input_var.flatten(2), mu, ls)
+                for mu, ls in zip(x_mu, x_ls))/L
+            prediction = T.sum(nn.layers.get_output(l_x_mu_list, deterministic=deterministic,
+                withNoise=withNoise), axis=0)/L
+        loss = -1 * (logpxz + kl_div)
+        return loss, prediction
+
+    # If there are dropout layers etc these functions return masked or non-masked expressions
+    # depending on if they will be used for training or validation/test err calcs
+    loss, _ = build_loss(deterministic=False)
+    test_loss, test_prediction = build_loss(deterministic=True, withNoise=False)
+
+    # functions for generating images given a code (used for visualization)
+    # for an given code z, we deterministically take x_mu as the generated data
+    # (no Gaussian noise is used to either encode or decode).
+    z_var = T.vector()
+    if binary:
+        generated_x = nn.layers.get_output(l_x, {l_z_mu:z_var}, withNoise=False,
+                deterministic=True)
+    else:
+        generated_x = nn.layers.get_output(l_x_mu_list[0], {l_z_mu:z_var}, withNoise=False,
+                deterministic=True)
+    gen_fn = theano.function([z_var], generated_x)
```

Comment:
For clarity, I'd move building the `gen_fn` function down into the `if z_dim == 2` case, because it's not used during training and only in that case. This reduces mental load while reading the code from top to bottom.